### PR TITLE
Fixed parsing of S3 URL

### LINKED ISSUE #10998: Issues when use S3 when using Amazon bucket
2 blocker issues with AWS s3:
1) can't parse one of the (proper) URL format (`Bucket name length out of bounds in S3 URI`)
i.e. (https://s3.us-east-2.amazonaws.com/bucketname/).
2) can't login to non default region `AuthorizationHeaderMalformed: the region 'XXXX' is wrong; expecting 'YYYY'`  (see https://github.com/aws/aws-sdk-cpp/issues/1057)

And 3rd (rather minor/cosmetic/need to be rechecked):
not showing the real error / reason of the issue here:
https://github.com/ClickHouse/ClickHouse/blob/f981649213fb519c57fda3dc2c51ab709cb327f4/src/Storages/StorageS3.cpp#L242-L248

### PR DESCRIPTION
**Pull request by excitoon**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Bug Fix

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Fixed parsing of S3 URLs.

Fixes: #10998

### CODE REVIEW

File: src/IO/S3Common.cpp [hunk 1]
```diff
@@ -90,8 +91,23 @@ namespace S3
         const String & secret_access_key)
     {
         Aws::Client::ClientConfiguration cfg;
+
         if (!endpoint.empty())
+        {
+            static const RE2 region_pattern(R"(s3\.([a-z0-9\-]+)\.amazonaws\.)");
```
Comment (qoega) [inline src/IO/S3Common.cpp:97]: Hard coded amazonaws seems not viable. There are many S3 compatible object storage providers

**Review by qoega — CHANGES_REQUESTED** [review]

File: src/IO/S3Common.cpp (hunk 1 above)
Comment (excitoon) [reply src/IO/S3Common.cpp:97]: I guess even for S3 it is not complete... How would you recommend to determine the region?

Comment (filimonov) [reply src/IO/S3Common.cpp:97]: It is used only for region detection. For now we don't know anything about requirements to pass/support region for non aws s3. 


region
Specifies the AWS region to use, such as us-east-1. By default, the region used is the default region configured in the applicable AWS credentials.

Comment (filimonov) [reply src/IO/S3Common.cpp:97]: I.e. If regexp will not match we just not pass the region, and everything works as before for non-aws

Comment (alexey-milovidov) [reply src/IO/S3Common.cpp]: Ok.

**Review by alexey-milovidov — APPROVED** [review]
Ok, but need to check test failures...

**Comment by alexey-milovidov** [comment]
> Caught exception while loading metadata: Code: 499, e.displayText() = DB::Exception: The specified bucket does not exist

Looks like already fixed in #11142

**Comment by alexey-milovidov** [comment]
Sorry, I did not notice that the test was broken. Will fix.

**Comment by alexey-milovidov** [comment]
It's strange that no test were added (how we are going to support it?)

**Comment by excitoon** [comment]
> It's strange that no test was added (how we are going to support it?)

Will be fixed in #11174

**Comment by excitoon** [comment]
@alexey-milovidov Sorry for that!

File: src/Storages/StorageS3.cpp [hunk 2]
```diff
@@ -244,7 +244,9 @@ Strings listFilesWithRegexpMatching(Aws::S3::S3Client & client, const S3::URI &
         {
             throw Exception("Could not list objects in bucket " + quoteString(request.GetBucket())
                     + " with prefix " + quoteString(request.GetPrefix())
-                    + ", page " + std::to_string(page), ErrorCodes::S3_ERROR);
+                    + ", page " + std::to_string(page)
```
Comment (filimonov) [inline src/Storages/StorageS3.cpp:247]: Dont print page if it's 1.

Comment (excitoon) [reply src/Storages/StorageS3.cpp:247]: Will fix that in #11174

**Pull request merged**